packages/jest-mock/__typetests__/mock-functions.test.ts

```diff
@@ -25,545 +20,665 @@ import {
+        .mockReset()
+        .mockImplementation(() => 'value')
+        .mockImplementationOnce(() => 'value')
+        .mockName('mock')
+        .mockReturnThis()
+        .mockReturnValue('value')
+        .mockReturnValueOnce('value'),
+    ).type.toEqual<Mock<() => string>>();
+
+    expect(
+      fn(() => 'value').mockReturnValue(Promise.resolve('value')),
+    ).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockReturnValueOnce(Promise.resolve('value')),
+    ).type.toRaiseError();
+  });
+
+  test('when async function is provided, returned object can be chained', () => {
+    expect(
+      fn(async () => 'value')
+        .mockClear()
+        .mockReset()
+        .mockImplementation(fn(async () => 'value'))
+        .mockImplementationOnce(fn(async () => 'value'))
+        .mockName('mock')
+        .mockResolvedValue('value')
+        .mockResolvedValueOnce('value')
+        .mockRejectedValue('error')
+        .mockRejectedValueOnce('error')
+        .mockReturnThis()
+        .mockReturnValue(Promise.resolve('value'))
+        .mockReturnValueOnce(Promise.resolve('value')),
+    ).type.toEqual<Mock<() => Promise<string>>>();
+
+    expect(fn(() => 'value').mockResolvedValue('value')).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockResolvedValueOnce('value'),
+    ).type.toRaiseError();
+
+    expect(fn(() => 'value').mockRejectedValue('error')).type.toRaiseError();
+    expect(
+      fn(() => 'value').mockRejectedValueOnce('error'),
+    ).type.toRaiseError();
+  });
+
+  test('models typings of mocked function', () => {
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    expect(fn()).type.toMatch<Function>();
+
+    expect(fn()).type.toEqual<Mock<(...args: Array<unknown>) => unknown>>();
+    expect(fn(() => {})).type.toEqual<Mock<() => void>>();
+    expect(fn((a: string, b?: number) => true)).type.toEqual<
+      Mock<(a: string, b?: number | undefined) => boolean>
+    >();
+    expect(
+      fn((e: any) => {
+        throw new Error();
+      }),
+    ).type.toEqual<Mock<(e: any) => never>>();
+
+    expect(fn('moduleName')).type.toRaiseError();
+  });
+
+  test('infers argument and return types of mocked function', () => {
+    expect(mockFn('one', 2)).type.toBeBoolean();
+    expect(mockAsyncFn(false)).type.toEqual<Promise<string>>();
+
+    expect(mockFn()).type.toRaiseError();
+    expect(mockAsyncFn()).type.toRaiseError();
+  });
+
+  test('infers argument and return types of mocked object', () => {
+    expect(new MockObject('credentials')).type.toEqual<{
+      connect(): Mock<(...args: Array<unknown>) => unknown>;
+      disconnect(): void;
+    }>();
+
+    expect(new MockObject()).type.toRaiseError();
+  });
+
+  test('.getMockImplementation()', () => {
+    expect(mockFn.getMockImplementation()).type.toEqual<
+      ((a: string, b?: number | undefined) => boolean) | undefined
+    >();
+
+    expect(mockFn.getMockImplementation('some-mock')).type.toRaiseError();
```

Comment:
`.toRaiseError()` check the error message, but I avoid these in this PR. 1. This is exactly what was tested before, so there is no regression. 2. The intention of this test is not about error message, the goal is to test this:

```ts
expect(mockFn.getMockImplementation).type.not.toBeCallableWith('some-mock');
``` 

The ability matchers, like matchers like `.toBeCallableWith()`, should ship with TSTyche 2. So I will use them here later.